demos/common/processing/async_pipeline.h

```diff
@@ -0,0 +1,143 @@
+    PipelineBase(std::unique_ptr<ModelBase> modelInstance, const CnnConfig& cnnConfig, InferenceEngine::Core* engine = nullptr);
+    virtual ~PipelineBase();
+
+    /// Waits until either output data becomes available or pipeline allows to submit more input data.
+    /// @param shouldKeepOrder if true, function will treat results as ready only if next sequential result (frame) is
+    /// ready (so results can be extracted in the same order as they were submitted). Otherwise, function will return if any result is ready.
+    void waitForData(bool shouldKeepOrder = true);
+
+    /// Returns true if there's available infer requests in the pool
+    /// and next frame can be submitted for processing.
+    /// @returns true if there's available infer requests in the pool
+    /// and next frame can be submitted for processing, false otherwise.
+    bool isReadyToProcess() { return requestsPool->isIdleRequestAvailable(); }
+
+    /// Returns performance info
+    /// @returns performance information structure
+    PerformanceInfo getPerformanceInfo() { std::lock_guard<std::mutex> lock(mtx); return perfInfo; }
+
+    /// Waits for all currently submitted requests to be completed.
+    ///
+    void waitForTotalCompletion() { if(requestsPool.get())requestsPool->waitForTotalCompletion(); }
+
+    /// Submit request to network
+    /// @param image - image to submit for processing
+    /// @returns -1 if image cannot be scheduled for processing (there's no any free InferRequest available).
+    /// Otherwise returns unique sequential frame ID for this particular request. Same frame ID will be written in the response structure.
+    virtual int64_t submitImage(cv::Mat img);
+
+    /// Gets available data from the queue 
+    /// @param shouldKeepOrder if true, function will treat results as ready only if next sequential result (frame) is
+    /// ready (so results can be extracted in the same order as they were submitted). Otherwise, function will return if any result is ready.
+    virtual std::unique_ptr<ResultBase> getResult(bool shouldKeepOrder=true);
+
+    /// Gets available data from the queue and renders it to output frame
+    /// This function should be overriden in inherited classes to provide default rendering of processed data
+    /// @returns rendered frame, its size corresponds to the size of network output
+
+protected:
+    /// Submit request to network
+    /// @param request - request to be submitted (caller function should obtain it using getIdleRequest)
+    /// @param metaData - additional source data. This is optional transparent data not used in inference process directly.
+    /// It is passed to inference result directly and can be used in postprocessing.
+    /// @returns unique sequential frame ID for this particular request. Same frame ID will be written in the responce structure.
+    virtual int64_t submitRequest(InferenceEngine::InferRequest::Ptr request,std::shared_ptr<MetaData> metaData);
```

Comment:
Take const refs as arguments.

Reply:
Const refs are corrected.

Reply:
No, we should not reorder fields as it is unsafe. Current approach gives safe solution for any fields order

Reply:
How is it unsafe?

Reply:
it will work with any members order. That's safer.
I can put it to both desturctors by the way, to make sure

Reply:
as we agreed, we leave it as is